### ISSUE TITLE
Support apt-mark hold

### ISSUE THREAD

**Issue by scottnonnenberg** [ansible/ansible-modules-core#19]
##### ISSUE TYPE

Feature Idea
##### COMPONENT NAME

apt
##### ANSIBLE VERSION

N/A
##### SUMMARY

This issue had reasonable interest, and all the `apt`-related modules are Core Modules, so opening a bug here. Here's the original issue: https://github.com/ansible/ansible/issues/8060

**Comment by sidharthsurana** [comment]
+1

**Comment by mpdehaan** [comment]
Can you please copy the description in to the new ticket?

Thanks!

**Comment by sidharthsurana** [comment]
## Summary from https://github.com/ansible/ansible/issues/8060

Support apt-mark hold/unhold to mark a package as ineligible for automatic updates. Otherwise, ansible can install a particular version of a package, only to see it blown away with a more recent one later.

https://help.ubuntu.com/community/PinningHowto#Introduction%20to%20Holding%20Packages

Steps To Reproduce:
- action: apt pkg=condor=8.0.7-250355 state=present
  ... later
- action: apt upgrade=dist
  ends up with an unexpected version of the package installed.

Expected Results:

Would like to be able to do something like this:
- action: apt pkg=condor=8.0.7-250355 state=present hold=yes
  hold=yes/no would correspond to apt-mark hold/unhold.

For completeness, you could support the manual/auto flag too. So another syntax would be
- action: apt pkg=condor=8.0.7-250355 state=present mark=hold
  ... or ...
- action: apt pkg=condor=8.0.7-250355 state=present mark=hold,manual
  Arguably, whenever you use ansible to install a specific version of a package then it should be held (and implicitly unheld if you specify "state=latest").

ansible 1.4 doesn't appear to do that, but maybe it's changed in later version.

Actual Results:

Currently no obvious ansible way to mark a package as held.

Workaround: use shell to run "apt-mark hold ". But this will be run on every ansible run, and relies on apt-mark itself being idempotent.

**Comment by bobobox** [comment]
This would be nice to have, especially as an older version of a package from a Linux repo probably can't easily be re-installed if inadvertently upgraded.

**Comment by candlerb** [comment]
Probably the right solution is as follows.
- Whenever you install a package with a specific version and `state=present`, the hold flag is automatically set
- Conversely, if you set `state=latest` then the hold flag is automatically cleared.

This involves no additional module options, but just "does the right thing".

Additional data point. At the moment, if the hold flag is set, and you specifically request a particular (different) version in a playbook, ansible aborts with the following error:

```
TASK: [htcondor | apt pkg=condor=8.2.4-281588 state=present] ******************
failed: [host.int.example.net] => {"failed": true}
stderr: E: There are problems and -y was used without --force-yes

stdout: Reading package lists...
Building dependency tree...
Reading state information...
The following packages were automatically installed and are no longer required:
  libaudio-scrobbler-perl libconfig-inifiles-perl linux-headers-3.2.0-32
  linux-headers-3.2.0-29 linux-headers-3.2.0-29-generic
  linux-headers-3.2.0-32-generic libssl0.9.8
Use 'apt-get autoremove' to remove them.
The following held packages will be changed:
  condor
The following packages will be upgraded:
  condor
1 upgraded, 0 newly installed, 0 to remove and 94 not upgraded.
Need to get 32.1 MB of archives.
After this operation, 4680 kB of additional disk space will be used.

msg: 'apt-get install 'condor=8.2.4-281588' ' failed: E: There are problems and -y was used without --force-yes


FATAL: all hosts have already failed -- aborting
```

So it looks like `--force-yes` needs to be provided by ansible if the package is held.

The playbook workaround I'm using for now is this ugly and non-idempotent approach:

```
- action:  shell apt-mark unhold condor
  ignore_errors: yes

- action: apt pkg=condor=8.2.4-281588 state=present

- action:  shell apt-mark hold condor
  ignore_errors: yes
```

**Comment by kosli** [comment]
+1

**Comment by jakommo** [comment]
+1

**Comment by LinusU** [comment]
@mpdehaan I would to have this implemented, would the behaviour outlined by @candlerb be accepted?

**Comment by LinusU** [comment]
Also, I think that `sudo apt-get install wine=1.2.3` will still work, hold will just prevent it from being upgraded with `dist-upgrade`, `upgrade`, etc...

**Comment by dmuysson** [comment]
+1
We are also using the workaround proposed by @candlerb but it's problematic. You always have to ignore errors as trying to hold an already held package throws errors (as does unholding a package which isn't held). 
Additionally, the other issue noted where --force fails to update a held package has halted more than one playbook in the past, leading to even more steps before an apt-update to try and unhold (potentially) held packages.

In short, supporting the ability to hold/unhold packages via the apt module would lead to cleaner playbooks, and the ability to upgrade held packages via --force would help prevent playbook failure.

**Comment by weevilgenius** [comment]
With versions of apt-mark which support the "showhold" command, you can make holds idempotent with something like this:

```
- command: apt-mark showhold
  register: held_packages
  changed_when: false

- command: apt-mark hold cassandra
  when: '"cassandra" not in held_packages.stdout'
```

**Comment by mfriedenhagen** [comment]
We use a different approach by specifying the version in `/etc/apt/preferences.d/fixed-casandra` like:

```
- copy:
    dest: /etc/apt/preferences.d/fixed-casandra
    content: |
      Package: casandra
      Pin: release v={{casandra_version}}
      Priority: 1002
- apt:
    name: cassandra
    state: latest
```

**Comment by dgellow** [comment]
@mfriedenhagen Thank you for your example. 
Just a correction, instead of 

```
      Pin: release v={{casandra_version}}
```

I had to use

```
      Pin: version {{casandra_version}}
```

**Comment by mfriedenhagen** [comment]
The good thing about the file solution: you may fixate the repository to use as well and may run an occasional apt-get distupgrade without hell breaking loose _and_ no command output has to be parsed.

**Comment by 2rs2ts** [comment]
Even if you don't support `apt-mark` in general it would still be nice to give the `apt` module something like the `yum` module's `exclude` option which could be implemented by using `apt-mark` behind the scenes.

**Comment by ansibot** [comment]
@ansible, ping.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by candlerb** [comment]
This issue was moved to ansible/ansible#18889
